sdk/subnetid.go

```diff
@@ -1,165 +1,211 @@
 package sdk
 
 import (
+	"bytes"
+	"hash/fnv"
 	"path"
-	"path/filepath"
+	"strconv"
 	"strings"
 
 	"github.com/filecoin-project/go-address"
 )
 
 type SubnetID struct {
-	Parent string
-	Actor  address.Address
+	// ChainID of the root network
+	Root uint64
+	// Children up to the current subnet.
+	Children []address.Address
 }
 
-var id0, _ = address.NewIDAddress(0)
-
 const (
-	RootStr          = "/root"
+	RootPrefix       = "r"
 	SubnetSeparator  = "/"
 	UndefStr         = ""
 	IPCAddrSeparator = ":"
-)
 
-// RootSubnet is the ID of the root network
-var RootSubnet = SubnetID{
-	Parent: RootStr,
-	Actor:  id0,
-}
+	// MaxChainID is the maximum chain ID value
+	// possible.
+	MaxChainID = 0xffffffffffffffff
+)
 
 // UndefSubnetID is the undef ID
 var UndefSubnetID = SubnetID{
-	Parent: SubnetSeparator,
-	Actor:  id0,
+	Root:     0,
+	Children: []address.Address{},
 }
 
 func (id SubnetID) Key() string {
-	return id.String()
+	b, _ := id.Bytes()
+	return string(b)
 }
 
-func NewSubnetIDFromString(addr string) (SubnetID, error) {
-	var out SubnetID
-	if addr == RootSubnet.String() {
-		out = RootSubnet
-		return out, nil
+// Equal checks if two subnet IDs are equal.
+func (id SubnetID) Equal(other SubnetID) bool {
+	if id.Root != other.Root {
+		return false
+	}
+	if len(id.Children) != len(other.Children) {
+		return false
 	}
-	dir, file := filepath.Split(addr)
-	act, err := address.NewFromString(file)
+	for i, c := range id.Children {
+		if c != other.Children[i] {
+			return false
+		}
+	}
+	return true
+}
+
+// IsRoot Returns true if the current subnet is the root subnet
+func (id SubnetID) IsRoot() bool {
+	return len(id.Children) == 0
+}
+
+func (id SubnetID) ChainID() uint64 {
+	if id.IsRoot() {
+		return id.Root
+	}
+
+	h := fnv.New64a()
+	h.Write([]byte(id.String()))
```

Comment:
Unhandled error. 
Can we just ignore it by adding `// nolint ` ?

Reply:
Good catch, updating it.